### PR TITLE
Move project metadata from setup.py to pyproject.toml

### PR DESCRIPTION
**Pull request by mwtoews**
This PR take's on [PEP 621](https://peps.python.org/pep-0621/) as supported by [setuptools](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html) to move the project's metadata from `setup.py` to `pyproject.toml`.

Dynamic features that are supported fetch the version from `keras.src.version.__version__` <s>and the common dependencies from `requirements-common.txt` (it cannot read `requirements.txt`, as only a subset of the requirements format is supported)</s>.

Modern projects shouldn't need `setup.py` anymore, so this file is removed. But if it is expected, a shim with `from setuptools import setup; setup()` can be added.

`pip_build.py` is updated to support the changes, as well as a few GHA workflows.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,5 +1,45 @@
...
+[tool.setuptools.dynamic]
+version = {attr = "keras.src.version.__version__"}
+dependencies = {file = "requirements-common.txt"}
+
+[tool.setuptools.packages.find]
+include = ["keras", "keras.*"]
+
 [tool.black]
 line-length = 80
+target-version = []
```
Comment (mwtoews) [inline pyproject.toml:42]: Normally black would get the target version from other project metadata (above). But this is overridden (for now) to stop black from formatting several other Python files with trivial edits. These possible edits can be evaluated in a separate PR.

File: pip_build.py [hunk 2]
```diff
@@ -22,29 +22,28 @@
...
+from pathlib import Path
 
 # Needed because importing torch after TF causes the runtime to crash
 import torch  # noqa: F401
 
 package = "keras"
 build_directory = "tmp_build_dir"
 dist_directory = "dist"
-to_copy = ["setup.py", "README.md"]
+to_copy = ["pyproject.toml", "README.md", "requirements-common.txt"]
```
Comment (fchollet) [inline pip_build.py:33]: Why copy the requirements file? To note, actual public-facing requirements are listed in setup.py. The `requirements-common.txt` file cannot be used on its own (it's incomplete) and the other requirements files are dev-facing (deps needed to run CI, not needed to run the package)

File: pyproject.toml [hunk 3]
```diff
@@ -1,5 +1,45 @@
...
+]
+dynamic = ["version", "dependencies"]
+
+[project.urls]
+Home = "https://keras.io/"
+Repository = "https://github.com/keras-team/keras"
+
+[tool.setuptools.dynamic]
+version = {attr = "keras.src.version.__version__"}
+dependencies = {file = "requirements-common.txt"}
```
Comment (fchollet) [inline pyproject.toml:35]: So, this is not the right set

File: pyproject.toml [hunk 4]
```diff
@@ -1,5 +1,45 @@
...
+
+[tool.setuptools.packages.find]
+include = ["keras", "keras.*"]
```
Comment (fchollet) [inline pyproject.toml:38]: No excludes?

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: pip_build.py (hunk 2 above)
Comment (mwtoews) [reply pip_build.py]: Ok, removed here as it doesn't seem to make any difference.

File: pyproject.toml (hunk 3 above)
Comment (mwtoews) [reply pyproject.toml:35]: This can be restored with the list from setup.py. But I thought sourcing this file would avoid duplication, as the former list is a subset of requirements-common.txt. Nevertheless, as documented, the user must meet the minimal install with `pip install -r requirements.txt`, which cannot be done with this setup due to the complexity of the requirements file. I've added a comment to re-iterate this detail.

File: pyproject.toml (hunk 4 above)
Comment (mwtoews) [reply pyproject.toml:39]: setup.py had `exclude=("*_test.py", "benchmarks")`, but this didn't do anything, so was dropped with the transition. The reason why it didn't do anything is that only Python modules under "keras" are found (thus "benchmarks" was already excluded), and individual Python files cannot be excluded -- only modules can be excluded. But I see that pip_build.py has logic to ignores the "_test.py" files.

File: pyproject.toml (hunk 3 above)
Comment (fchollet) [reply pyproject.toml:35]: We only want to install as few deps as possible on the user side, so we insert the explicit list here

**Review by fchollet — APPROVED** [review]
LGTM, thank you!
